geomstats/stratified_geometry/wald_space.py

```diff
@@ -0,0 +1,1024 @@
+
+        A list of dictionaries, each dictionary is for the respective connected
+        component of the forest, and the items of each dictionary are for each pair
+        of labels u, v, u < v in the respective component, a list of the splits on the
+        unique path between the labels u and v.
+
+        Returns
+        -------
+        label_paths : list of dict
+            The list of dictionaries containing list of splits on path between labels.
+        """
+        if self._paths is not None:
+            return self._paths
+        self._paths = [
+            {
+                (u, v): [s for s in splits if s.separates(u, v)]
+                for u, v in it.combinations(part, r=2)
+            }
+            for part, splits in zip(self.partition, self.split_sets)
+        ]
+        return self._paths
+
+    def corr(self, x):
+        """Compute the correlation matrix of the topology with edge weights ``x``.
+
+        Parameters
+        ----------
+        x : array-like
+            Takes a vector of length 'number of total splits' of the structure.
+
+        Returns
+        -------
+        corr : array-like, shape=[n, n]
+            Returns the corresponding correlation matrix.
+        """
+        corr = np.zeros((self.n, self.n))
+        for path_dict in self.paths():
+            for (u, v), path in path_dict.items():
+                corr[u][v] = np.prod([1 - x[self.where[split]] for split in path])
+                corr[v][u] = corr[u][v]
+        np.fill_diagonal(corr, 1)
+        return gs.array(corr, dtype=gs.float32)
+
+    def support(self):
+        r"""For each split, return a boolean matrix with entries indicating separation.
+
+        For each split, give an :math:`n\times n` dimensional matrix, where the
+        uv-th entry is ``True`` if the split separates the labels u and v, else
+        ``False``.
+
+        Returns
+        -------
+        support : list of array-like
+            The support of each split according to the topology.
+        """
+        if self._support is not None:
+            return self._support
+        _support = [
+            np.zeros((self.n, self.n), dtype=bool)
+            for _ in self.flatten(self.split_sets)
+        ]
+        for path_dict in self.paths():
+            for (u, v), path in path_dict.items():
+                for split in path:
+                    _support[self.where[split]][u, v] = True
+                    _support[self.where[split]][v, u] = True
+        self._support = [gs.array(m) for m in self.flatten(_support)]
+        return self._support
+
+    def corr_gradient(self, x):
+        """Compute the gradient of the correlation matrix, differentiated by x.
+
+        Parameters
+        ----------
+        x : array-like
+            The vector x at which the gradient is computed.
+
+        Returns
+        -------
+        gradient : list of array-like
```

Comment:
Why list of array like instead of array-like? What are the advantages?

Reply:
There are no particular advantages except that I know what a list of arrays does; I was not sure how gs.array([list of arrays]) would behave. That's why I've left it as a list for now. Should I do something like gs.array([list of arrays], axis=-1), or can I safely do gs.array([list of arrays]) and I'll get a three dimensional array?

Reply:
`gs.array([list of arrays])` will return a 3d array. Same behavior as `np.array`.